feat(zkstack): Split chain and ecosystem layers

matias-gonz opened the pull request:
## What ❔
Move commands into chain subcommand:
```
zkstack server -> zkstack chain server
zkstack consensus -> zkstack chain consensus
zkstack contract-verifier -> zkstack chain contract-verifier
```

Now it is possible to run `zkstack chain create` without having an ecosystem.
Also run `zkstack chain init` without ecosystem by providing a path to l1 wallets and l1 contracts configs.

matias-gonz committed:
Migrate server to chain

matias-gonz committed:
Migrate contract verifier to chain

matias-gonz committed:
Migrate consensus to chain

matias-gonz committed:
fmt

matias-gonz committed:
Fix consensus proto

matias-gonz committed:
Fix consensus command

matias-gonz committed:
Add ZkStackConfig enum

matias-gonz committed:
fmt

matias-gonz committed:
Refactor deploy paymaster

matias-gonz committed:
Refactor chain commands

matias-gonz committed:
rollback consensus

matias-gonz committed:
Refactor chain consensus setup

matias-gonz committed:
Merge branch 'main' into matias/refactor-3-layered-chain

matias-gonz committed:
Merge branch 'main' of github.com:matter-labs/zksync-era into matias/refactor-chain-commands

matias-gonz committed:
Update tests

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
fmt

matias-gonz committed:
Merge branch 'matias/refactor-chain-commands' of github.com:matter-labs/zksync-era into matias/refactor-chain-commands

matias-gonz committed:
fmt

matias-gonz committed:
generate autocompletion

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
lint autocomplete

matias-gonz committed:
generate autocomplete

matias-gonz committed:
lint

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
Fix ci

matias-gonz committed:
Make contract_verifier:init:run async

matias-gonz commented on zkstack_cli/crates/zkstack/src/commands/chain/contract_verifier/mod.rs:
```diff
@@ -17,6 +17,6 @@ pub enum ContractVerifierCommands {
     match args {
         ContractVerifierCommands::Run => run::run(shell).await,
-        ContractVerifierCommands::Init(args) => init::run(shell, args).await,
+        ContractVerifierCommands::Init(args) => init::run(shell, args),
```
https://github.com/matter-labs/zksync-era/pull/3199/commits/942e335d71d87e791e8de377f329660225af46b0

matias-gonz committed:
Remove use messages:self

matias-gonz commented on zkstack_cli/crates/zkstack/src/commands/chain/consensus/mod.rs:
```diff
@@ -18,7 +19,10 @@ use xshell::Shell;
 
-use crate::{messages, utils::consensus::parse_attester_committee};
+use crate::{
+    messages::{self},
```
https://github.com/matter-labs/zksync-era/pull/3199/commits/1a6f628cfd2e27bad0ee2bab6f2bf72ac2f35e11

matias-gonz committed:
Remove _zkstack.zsh

matias-gonz commented on zkstack_cli/_zkstack.zsh:
https://github.com/matter-labs/zksync-era/pull/3199/commits/f6fb8bfdb4c44d2a5d086e3a8f0123f0fc2b3c6c

matias-gonz committed:
Load chain in chain:run and pass it to the next commands

matias-gonz commented on zkstack_cli/crates/zkstack/src/commands/chain/contract_verifier/run.rs:
```diff
@@ -1,17 +1,15 @@
-    let chain = ecosystem
-        .load_current_chain()
-        .context(MSG_CHAIN_NOT_FOUND_ERR)?;
+    let chain = ZkStackConfig::load_current_chain(shell).context(MSG_CHAIN_NOT_INITIALIZED)?;
```
https://github.com/matter-labs/zksync-era/pull/3199/commits/a01eea872295eb73484a058f7dcf9642419ac8ef

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
Add ChainConfig::from_internal

matias-gonz commented on zkstack_cli/crates/config/src/zkstack_config.rs:
```diff
@@ -0,0 +1,56 @@
+            let artifacts = chain_internal
+                .artifacts_path
+                .context("Artifacts path not found")?;
+
```
https://github.com/matter-labs/zksync-era/pull/3199/commits/18403f15ed4f4b67dd93cb6064a6c0435c24c972

Deniallugo commented on zkstack_cli/crates/config/src/chain.rs:
```diff
@@ -156,13 +161,69 @@ impl ChainConfig {
         }
     }
+
+    pub fn from_internal(
```
does pub necessary?

Deniallugo replied:
pub(crate) is better

Deniallugo commented on zkstack_cli/crates/config/src/zkstack_config.rs:
```diff
@@ -0,0 +1,27 @@
+        }
+    }
+
+    pub fn load_current_chain(shell: &Shell) -> anyhow::Result<ChainConfig> {
```
let's name it just `current_chain` and also add a method for getting ecosystem

matias-gonz replied:
[bf1f8da](https://github.com/matter-labs/zksync-era/pull/3199/commits/bf1f8da46320e9e4cc98ed9ff6d8d4a91601bf11) & [b2d7388](https://github.com/matter-labs/zksync-era/pull/3199/commits/b2d73882587fdd60e3505e397aeb1c13b89f3878)

Deniallugo commented on zkstack_cli/crates/config/src/ecosystem.rs:
let's remove `pub` from     `pub fn from_file(shell: &Shell) -> Result<Self, EcosystemConfigFromFileError>`

In this case it'd be impossible to deserialize config outside ZkstackConfig

matias-gonz replied:
[6406717](https://github.com/matter-labs/zksync-era/pull/3199/commits/6406717e65db22305e6f3dc13b06d35f1842db39)

Deniallugo commented on zkstack_cli/crates/config/src/chain.rs:
```diff
@@ -156,13 +161,69 @@ impl ChainConfig {
+}
+
+impl ChainConfigInternal {
+    pub fn from_file(shell: &Shell) -> anyhow::Result<ChainConfigInternal> {
```
let's remove `pub`

matias-gonz replied:
[2c5cf58](https://github.com/matter-labs/zksync-era/pull/3199/commits/2c5cf589dd23062ab22aefaac4e1d7028892dd21)

matias-gonz committed:
Rename ZkStackConfig::load_current_chain to current_chain

matias-gonz committed:
Add ZkStackConfig::ecosystem

matias-gonz committed:
Make EcosystemConfig::from_file private

matias-gonz committed:
Make ChainConfigInternal::from_file private

matias-gonz committed:
Make ChainConfig::from_internal private

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
Merge branch 'main' of github.com:matter-labs/zksync-era into matias/refactor-chain-commands

matias-gonz committed:
fmt

matias-gonz committed:
Add autocompletion files

matias-gonz committed:
Add zsh autocomplete file

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
Fix ci

matias-gonz committed:
Fix contract-verifier ci

matias-gonz committed:
Fix contract-verifier ci

Deniallugo commented on zkstack_cli/crates/zkstack/src/commands/chain/init/configs.rs:
```diff
@@ -26,7 +26,7 @@ use crate::{
 
 pub async fn run(args: InitConfigsArgs, shell: &Shell) -> anyhow::Result<()> {
-    let ecosystem_config = EcosystemConfig::from_file(shell)?;
+    let ecosystem_config = ZkStackConfig::ecosystem(shell)?;
```
we shouldn't use it here. 
We need it only for port reallocation. 
It'd be better to pass zkstack config here and if  ecosystem is available do reallocation and if not, use default one

matias-gonz replied:
We also need ecosystem for contracts config, this is fixed in a next PR:
https://github.com/matter-labs/zksync-era/pull/3231

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
Merge branch 'main' of github.com:matter-labs/zksync-era into matias/refactor-chain-commands

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

matias-gonz committed:
feat(zk_toolbox): Make chain create independent from ecosystem (#3210)

## What ❔

<!-- What are the changes this PR brings about? -->
<!-- Example: This PR adds a PR template to the repo. -->
<!-- (For bigger PRs adding more context is appreciated) -->

## Why ❔

<!-- Why are these changes done? What goal do they contribute to? What
are the principles behind them? -->
<!-- Example: PR templates ensure PR reviewers, observers, and future
iterators are in context about the evolution of repos. -->

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [ ] PR title corresponds to the body of PR (we generate changelog
entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zkstack dev fmt` and `zkstack dev
lint`.

matias-gonz committed:
feat(zk_toolbox): Make chain init independent from ecosystem (#3231)

## What ❔

<!-- What are the changes this PR brings about? -->
<!-- Example: This PR adds a PR template to the repo. -->
<!-- (For bigger PRs adding more context is appreciated) -->

## Why ❔

<!-- Why are these changes done? What goal do they contribute to? What
are the principles behind them? -->
<!-- Example: PR templates ensure PR reviewers, observers, and future
iterators are in context about the evolution of repos. -->

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [ ] PR title corresponds to the body of PR (we generate changelog
entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zkstack dev fmt` and `zkstack dev
lint`.

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

sanekmelnikov reviewed (commented):
@matias-gonz, would it be reasonable to ask you to fill in the description [What, Why]? :)
Since it's a breaking change, I think it would be useful to list all the commands that will stop working (`zkstack server` -> `zkstack chain server`, etс.)

@deniallugo, @matias-gonz
So I've been trying out these changes and I guess I don't understand what problem are we trying to solve or what use-case are we trying to improve?
1) If we want to deploy a brand-new ecosystem, I don't see much value separating ecosystem and chain. imho, the current setup works really well for that case.
2) If we are working with the ecosystem, that already exists:
2.1) if ecocsystem is on localhost: I don't see how `zkstack chain create/init` is easier to use outside of the ecosystem - because I need to provide contracts/wallets file and path to code. It's much easier to run the same commands from the ecosystem folder.
2.2) if ecosystem is ZKSync's Sepolia/Mainnet: `zkstack ecosystem import --from mainnet` in my view would be better way to go since contracts/wallet (addresses) can be populated for you
2.3) if ecosystem is not ZKSync's Sepolia/Mainnet: that's probably the only use-case that will be improved, although I'd still prefer `zkstack ecosystem import --from [FILE]` followed by `zkstack chain create` etc. and also that's the least common use-case.
I think the fundamental problem is that we have ecosystem and chain command tightly coupled together, but making it easy to import existing ecosystem with a separate command is a more straightforward way to go?

The biggest problem with this PR is that all of these currently aren't working for chain-only folder:
* `zkstack portal`
* `zkstack explorer`
* `zkstack update`
* `zkstack external-node`
* `zkstack containers`
* `zkstack prover`

As well as:
* `zkstack chain init configs`

As for `zkstack chain server` vs `zkstack server`: do we really think it's worth breaking for all the users? For chain-only folders `zkstack chain server` makes less sense than `zkstack server` since obviously everything we're going to run will be chain-related. Can we support both?

sanekmelnikov commented on zkstack_cli/crates/zkstack/src/commands/chain/create.rs:
```diff
@@ -21,22 +22,49 @@ use crate::{
+        .map(|ecosystem| ecosystem.list_of_chains().len() as u32)
+        .unwrap_or(0);
+
+    let internal_id = ecosystem.as_ref().map_or(0, |_| number_of_chains + 1);
```
Also, is it used anywhere apart from calculating ports?

matias-gonz replied:
https://github.com/matter-labs/zksync-era/pull/3210#discussion_r1836182204
We decided to keep it as 0

sanekmelnikov commented:
Ah, with chain registrar changes though, I can see the value for external chains.
You basically will be able to:
`zkstack chain create`
`zkstack chain init configs` (don't even need to pass the ecosystem wallets)
`zkstack chain registration send-proposal`
`zkstack chain accept-ownership`
and you are good to go.

Deniallugo commented on zkstack_cli/crates/zkstack/src/commands/chain/create.rs:
```diff
@@ -85,16 +112,18 @@ pub(crate) fn create_chain_inner(
     if args.evm_emulator && !has_evm_emulation_support {
         anyhow::bail!(MSG_EVM_EMULATOR_HASH_MISSING_ERR);
     }
+    let rocks_db_path = chain_path.join(LOCAL_DB_PATH);
+    let artifacts = chain_path.join(LOCAL_ARTIFACTS_PATH);
```
@matias-gonz please check it

matias-gonz replied:
Yes, I introduced this bug, working on it!

matias-gonz replied:
[8584649](https://github.com/matter-labs/zksync-era/pull/3199/commits/85846499834599964584d0df78fe711da658a848)

matias-gonz committed:
Fix conflicts

matias-gonz committed:
feat(zkstack): Run containers with either chain or ecosystem (#3345)

## What ❔

<!-- What are the changes this PR brings about? -->
<!-- Example: This PR adds a PR template to the repo. -->
<!-- (For bigger PRs adding more context is appreciated) -->

## Why ❔

<!-- Why are these changes done? What goal do they contribute to? What
are the principles behind them? -->
<!-- Example: PR templates ensure PR reviewers, observers, and future
iterators are in context about the evolution of repos. -->

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [ ] PR title corresponds to the body of PR (we generate changelog
entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zkstack dev fmt` and `zkstack dev
lint`.

matias-gonz committed:
fix(zkstack): Fix duplicated chain layer (#3348)

## What ❔

<!-- What are the changes this PR brings about? -->
<!-- Example: This PR adds a PR template to the repo. -->
<!-- (For bigger PRs adding more context is appreciated) -->

## Why ❔

<!-- Why are these changes done? What goal do they contribute to? What
are the principles behind them? -->
<!-- Example: PR templates ensure PR reviewers, observers, and future
iterators are in context about the evolution of repos. -->

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [ ] PR title corresponds to the body of PR (we generate changelog
entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zkstack dev fmt` and `zkstack dev
lint`.

matias-gonz committed:
Rename ecosystem_wallets_path

matias-gonz committed:
Lint

matias-gonz commented on zkstack_cli/crates/zkstack/src/commands/chain/args/init/mod.rs:
```diff
@@ -35,6 +39,10 @@ pub struct InitArgs {
+    #[clap(long, help = MSG_ECOSYSTEM_CONTRACTS_PATH_HELP)]
+    pub ecosystem_contracts_path: Option<String>,
+    #[clap(long, help = MSG_WALLETS_PATH_HELP)]
+    pub wallets_path: Option<String>,
```
[feaae79](https://github.com/matter-labs/zksync-era/pull/3199/commits/feaae7946b6dc85c02f64a65b8e0d85469d82382)

matias-gonz committed:
Merge branch 'main' into matias/refactor-chain-commands

deniallugo-ml commented:
Closing this PR as the `zkstack` CLI was substantially restructured after this PR was opened. The code paths, command structure, and configuration handling this PR targets have all changed, so the diff no longer applies to the current codebase.

If the underlying feature or fix is still needed, please open a new PR against the current CLI structure.
